metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1123,29 +1165,104 @@ private void claim(int epoch) {
             // Prepend the activate event. It is important that this event go at the beginning
             // of the queue rather than the end (hence prepend rather than append). It's also
             // important not to use prepend for anything else, to preserve the ordering here.
-            queue.prepend(new ControllerWriteEvent<>("completeActivation[" + epoch + "]",
-                    new CompleteActivationEvent()));
+            ControllerWriteEvent<Void> activationEvent = new ControllerWriteEvent<>("completeActivation[" + epoch + "]",
+                    new CompleteActivationEvent(),
+                    EnumSet.of(DOES_NOT_UPDATE_QUEUE_TIME, RUNS_IN_PREMIGRATION));
+            activationEvent.future.whenComplete((__, t) -> {
+                if (t != null) {
+                    fatalFaultHandler.handleFault("exception while activating controller", t);
+                }
+            });
+            queue.prepend(activationEvent);
         } catch (Throwable e) {
             fatalFaultHandler.handleFault("exception while claiming leadership", e);
         }
     }
 
+    /**
+     * Generate the set of activation records. Until KIP-868 transactions are supported, these records
+     * are committed to the log as an atomic batch. The records will include the bootstrap metadata records
+     * (including the bootstrap "metadata.version") and may include a ZK migration record.
+     */
+    public static void generateActivationRecords(
```

Comment:
Can we just return a `List<ApiMessageAndVersion>`? I don't think we're creating enough records to make a new list an efficiency concern...